### ISSUE TITLE
Bug: Incorrect error for overload with number index signature

### ISSUE THREAD

**Issue by osi-oswald** [Microsoft/TypeScript#11885]
The following code results in an inconsistent compiler error:

**TypeScript Version:**  2.0 (as in [Playground](http://www.typescriptlang.org/play/#src=export%20function%20qwer%3CV%3E%28arg%3A%20%7B%5Bkey%3A%20number%5D%3A%20V%7D%29%3A%20V%5B%5D%0D%0Aexport%20function%20qwer%3CT%3E%28arg%3A%20T%29%3A%20T%0D%0Aexport%20function%20qwer%28arg%29%3A%20any%20%7B%0D%0A%7D%0D%0A%0D%0Atype%20Foo%20%3D%20%7Bbar%3A%20number%7D%3B%0D%0Avar%20foo%20%3D%20%7Bbar%3A%2042%7D%3B%0D%0Aqwer%28foo%29%20as%20Foo%3B%20%2F%2F%20compiler%20ERROR%0D%0Aqwer%28%7Bbar%3A%2042%7D%29%20as%20Foo%3B%20%2F%2F%20OK))

**[Code](http://www.typescriptlang.org/play/#src=export%20function%20qwer%3CV%3E%28arg%3A%20%7B%5Bkey%3A%20number%5D%3A%20V%7D%29%3A%20V%5B%5D%0D%0Aexport%20function%20qwer%3CT%3E%28arg%3A%20T%29%3A%20T%0D%0Aexport%20function%20qwer%28arg%29%3A%20any%20%7B%0D%0A%7D%0D%0A%0D%0Atype%20Foo%20%3D%20%7Bbar%3A%20number%7D%3B%0D%0Avar%20foo%20%3D%20%7Bbar%3A%2042%7D%3B%0D%0Aqwer%28foo%29%20as%20Foo%3B%20%2F%2F%20compiler%20ERROR%0D%0Aqwer%28%7Bbar%3A%2042%7D%29%20as%20Foo%3B%20%2F%2F%20OK)**

``` ts
export function qwer<V>(arg: {[key: number]: V}): V[]
export function qwer<T>(arg: T): T
export function qwer(arg): any {
}

type Foo = {bar: number};
var foo = {bar: 42};
qwer(foo) as Foo; // <-- compiler ERROR
qwer({bar: 42}) as Foo; // OK
```

**Expected behavior:**
Both `quer()` calls should compile.

**Actual behavior:** 
`quer(foo) as Foo` fails, despite `foo` is of the same type as object `{bar: 42}`,

**Comment by sandersn** [comment]
This difference is caused by contextual typing. `{ bar: 42 }` is contextually typed, and I think that it gets a number index from its contextual type -- which in this case is `{[key: number]: V}`.

However, I don't understand the error message for `foo`. I'll keep looking at it.

Here's the error I got:

```ts
Type '[]{}' cannot be converted to type '{ bar: number }'.
  Property 'bar' is missing in type '[]{}'.
```

**Comment by sandersn** [comment]
OK, the error is intended for the first call `qwer(foo)`. From reading the spec and the code, here's why:

The root `foo` has an object literal type `{ bar: number }`. You get the expected behaviour if you have

```ts
interface Foo { bar: number }
var foo: Foo = { bar: 42 };
```

That's because the compiler special cases object literals because they are in some sense more "open" than a class type or even an interface type. An object literal type is assignable to a type with an numeric index signature even if it doesn't have one. The compiler just checks that all the numeric properties match the index type (which it trivially does, for `foo`, since it has no numeric properties).

So the first overload matches. In the process, `V` gets bound to the type of `foo`'s numeric properties. Since it has none, `V={}`. `{}` is the type that the type parameter inference algorithm gives to types it can't find an answer for.

I'm not sure yet why the literal `{ bar: 42 }` doesn't fail in the same way. I'll look at that next.

**Comment by sandersn** [comment]
The reason that `{ bar: 42 }` fails to match the first overload is not contextual typing like I thought. It's actually excess property checking.  When an object literal is passed to a object-typed parameter, like `{ [key: number] : V }`, one of the first assignability checks is whether the literal has any properties that are not specified [1]. In this case `{ [key: number]: V }` does have not have *any* explicit properties, so `bar` is an error and the first overload fails.

So, confusingly, that means both behaviours are intended.

[1] This catches typos when passing property bags to a function (for example), but does not apply to object literals that were previously saved in a variable. Like `foo`. I think we decided it would break too much existing code, but I wasn't here for that decision.
